src/coprocessor/dag/rpn_expr/types/expr_eval.rs

```diff
@@ -129,19 +129,70 @@ impl RpnExpression {
         schema: &'a [FieldType],
         columns: &'a mut LazyBatchColumnVec,
     ) -> Result<RpnStackNode<'a>> {
-        assert!(rows > 0);
-
-        let mut stack = Vec::with_capacity(self.len());
-
         // We iterate two times. The first time we decode all referred columns. The second time
         // we evaluate. This is to make Rust's borrow checker happy because there will be
         // mutable reference during the first iteration and we can't keep these references.
+        self.ensure_columns_decoded(context, schema, columns)?;
+        self.eval_unchecked(context, rows, schema, columns)
+    }
+
+    /// Evaluates the expression into a boolean vector.
+    ///
+    /// # Panics
+    ///
+    /// Panics if referenced columns are not decoded.
+    ///
+    /// Panics if the boolean vector output buffer is not large enough to contain all values.
+    pub fn eval_as_mysql_bools(
+        &self,
+        context: &mut EvalContext,
+        rows: usize,
+        schema: &[FieldType],
+        columns: &mut LazyBatchColumnVec,
+        outputs: &mut [bool], // modify an existing buffer to avoid repeated allocation
+    ) -> Result<()> {
+        use crate::coprocessor::codec::data_type::AsMySQLBool;
 
+        assert!(outputs.len() >= rows);
+        let values = self.eval(context, rows, schema, columns)?;
+        match values {
+            RpnStackNode::Scalar { value, .. } => {
+                let b = value.as_mysql_bool(context)?;
+                for i in 0..rows {
+                    outputs[i] = b;
+                }
+            }
+            RpnStackNode::Vector { value, .. } => {
+                assert_eq!(value.len(), rows);
+                value.eval_as_mysql_bools(context, outputs)?;
+            }
+        }
+        Ok(())
+    }
+
+    pub fn ensure_columns_decoded<'a>(
+        &'a self,
+        context: &mut EvalContext,
+        schema: &'a [FieldType],
+        columns: &'a mut LazyBatchColumnVec,
+    ) -> Result<()> {
         for node in self.as_ref() {
             if let RpnExpressionNode::ColumnRef { ref offset, .. } = node {
                 columns.ensure_column_decoded(*offset, &context.cfg.tz, &schema[*offset])?;
             }
         }
+        Ok(())
+    }
+
+    pub fn eval_unchecked<'a>(
```

Comment:
This function is safe. It will just panic when we need a decoded column while it is not.
The `unchecked` name may be misleading. Need some suggestions as well...

Reply:
Actually I like this name pretty much 🤣 It's only problem is that it is not *that* "unsafe".